File: src/lib/api/planner.ts
/**
 * Study Planner API Client
 * ========================
 * Calls backend Study Planner Agent endpoints to generate and fetch personalized study schedules.
 */

export interface StudyBlock {
  day: string;
  topic: string;
  duration_minutes: number;
  priority: "high" | "normal" | "low";
  description: string;
  video_url?: string;
  video_query?: string;
}

export interface StudyPlan {
  id: string;
  student_id: string;
  title: string;
  summary: string;
  schedule: StudyBlock[];
  action_items: string[];
  created_at: string;
}

import { apiFetch, API_BASE_URL } from "./client";

/**
 * Fetches existing study schedules for a given student.
 */
export async function getStudentStudyPlans(studentId: string, token?: string): Promise<StudyPlan[]> {
  const headers: Record<string, string> = {};
  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  const response = await apiFetch(`${API_BASE_URL}/ai/planner/${encodeURIComponent(studentId)}`, {
    method: "GET",
    headers,
  });

  if (!response.ok) {
    throw new Error(`Status ${response.status}`);
  }

  return await response.json();
}

/**
 * Invokes the AI Planner Agent to generate a new custom study plan.
 */
export async function generateStudyPlan(
  studentId: string,
  targetDays = 5,
  customGoals?: string,
  availableHours = 2.0,
  learningSpeed = "moderate",
  token?: string
): Promise<StudyPlan> {
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
  };
  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  const response = await apiFetch(`${API_BASE_URL}/ai/planner/generate`, {
    method: "POST",
    headers,
    body: JSON.stringify({
      student_id: studentId || "demo_student",
      target_days: targetDays,
      custom_goals: customGoals || null,
      available_hours: availableHours,
      learning_speed: learningSpeed,
    }),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({ detail: "Generation error" }));
    throw new Error(errorData.detail || `Server returned status ${response.status}`);
  }

  const data = await response.json();
  return {
    id: data.plan_id,
    student_id: studentId,
    title: data.title,
    summary: data.summary,
    schedule: data.schedule,
    action_items: data.action_items,
    created_at: data.created_at,
  };
}

File: reference-backend/app/api/routes/planner.py
"""
Study Planner Routes
====================

Exposes endpoints for generating personalized AI study plans and retrieving
saved study schedules.
"""

import json
import logging
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session as DBSession

from app.api.deps import get_current_user, resolve_student_id, ensure_owns_student
from app.db.session import get_db
from app.models.models import User, StudyPlan
from app.schemas.schemas import (
    PlannerGenerateRequest,
    PlannerGenerateResponse,
    StudyPlanResponse,
    StudyBlock,
)
from app.ai.agents.planner_agent import PlannerAgent

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/ai", tags=["planner"])

planner_agent = PlannerAgent()


@router.post("/planner/generate", response_model=PlannerGenerateResponse)
async def generate_study_plan(
    payload: PlannerGenerateRequest,
    db: DBSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> PlannerGenerateResponse:
    """
    Generates a personalized AI study schedule incorporating weak concepts and deadlines.
    """
    student_id = resolve_student_id(payload.student_id, current_user)

    try:
        result = await planner_agent.generate_plan(
            db=db,
            student_id=student_id,
            target_days=payload.target_days,
            custom_goals=payload.custom_goals,
            available_hours=payload.available_hours,
            learning_speed=payload.learning_speed,
        )

        schedule_blocks = [StudyBlock(**block) for block in result["schedule"]]

        return PlannerGenerateResponse(
            plan_id=result["plan_id"],
            title=result["title"],
            summary=result["summary"],
            schedule=schedule_blocks,
            action_items=result["action_items"],
            created_at=result["created_at"],
        )
    except Exception as e:
        logger.error("Study plan generation failed: %s", str(e), exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Study Planner Agent error: {str(e)}",
        )


@router.get("/planner/{student_id}", response_model=list[StudyPlanResponse])
def get_student_study_plans(
    student_id: str,
    db: DBSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[StudyPlanResponse]:
    """
    Retrieves saved study schedules for a given student.
    """
    ensure_owns_student(student_id, current_user)

    plans = (
        db.query(StudyPlan)
        .filter(StudyPlan.student_id == student_id)
        .order_by(StudyPlan.created_at.desc())
        .limit(5)
        .all()
    )

    if not plans:
        # Honest empty result — no fabricated default plan.
        return []

    response = []
    for plan in plans:
        try:
            schedule_list = json.loads(plan.schedule_json)
        except Exception:
            schedule_list = []

        try:
            action_items_list = json.loads(plan.action_items_json)
        except Exception:
            action_items_list = []

        blocks = [StudyBlock(**b) for b in schedule_list if isinstance(b, dict)]

        response.append(
            StudyPlanResponse(
                id=plan.id,
                student_id=plan.student_id,
                title=plan.title,
                summary=plan.summary or "",
                schedule=blocks,
                action_items=action_items_list,
                created_at=plan.created_at,
            )
        )

    return response
